Add function to simulate keystrokes.

amosbird opened the pull request:
This patch is a rebase of the keysym branches, with some improvements and tenderly care of @aidanholm :)

karottenreibe committed:
added function to simulate keystrokes on the window

karottenreibe committed:
addded some guards for bad keys

karottenreibe committed:
also converting unicode chars like ":" etc.

karottenreibe committed:
added lua lib to parse keystrings

karottenreibe committed:
send_key now returns true/false based on sending success instead of
issuing a console warning

karottenreibe committed:
fixed minor bug in keystring sending

karottenreibe committed:
moved keysym.lua out of lousy and fixed some bugs

*   it needs the window object, so it can't be in lousy
*   fixed some typos

karottenreibe committed:
added the ability to send modifiers as well

karottenreibe committed:
added modifier shortcuts like S for Shift or M3 for Mod3

karottenreibe committed:
added modifier handling on the C side and fixed modifiers on the lua
side

mason-larobina committed:
Validate utf8 input to the window send_Key method.

mason-larobina committed:
Check gdk_keymap_get_entries_for_keyval return value.

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,71 @@
+--- Keysym for luakit.
```
I'm not happy with the name; should be "sendkeys" or "fakekeys" or something, but "keysym" is totally opaque

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,71 @@
+-- See gdk/gdkkeysyms.h for a complete list of recognized key names.
+--
+-- @module keysym
+-- Simulating keystrokes
```
remove extra line "Simulating keystrokes" here

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,71 @@
+-- them to the window. A keystring is a string of keys to press, with
+-- special keys denoted in between angle brackets:
+--
+--  <code>keysym.send(w, "<S-Home><BackSpace>")</code>
```
indent code with four spaces instead of using `<code>`

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,71 @@
+            symbol = ""
+        elseif char == ">" and symbol then
+            if #symbol == 0 then
+                w:error("bad keystring: " .. keystring)
```
- these are script errors, not user-facing error messages: use `error()` instead
- error message needs to say what's wrong with the keystring

a better version would be `error("keystring '" .. keystring .. "' contains empty modifier group")`; note the string concatenation; unfortunately `error()` doesn't support format strings (might hack that in at some point..)

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,71 @@
+                table.insert(modifiers, symbol)
+                symbol = ""
+            else
+                w:error("bad modifier in keystring: " .. symbol)
```
ditto

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,71 @@
+        end
+    end
+    if symbol then
+        w:error("bad keystring: " .. keystring)
```
ditto

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,71 @@
+    end
+    for _, key in ipairs(keys) do
+        if not w.win:send_key(key.key, key.mods) then
+            w:error("failed to send keystroke: " .. key.key)
```
ditto

aidanholm commented on lib/window.lua:
```diff
@@ -166,7 +166,11 @@ local init_funcs = {
 
     key_press_match = function (w)
-        w.win:add_signal("key-press", function (_, mods, key)
+        w.win:add_signal("key-press", function (_, mods, key, send)
```
`send` should be `synthetic`

aidanholm commented on lib/window.lua:
```diff
@@ -166,7 +166,11 @@ local init_funcs = {
-        w.win:add_signal("key-press", function (_, mods, key)
+        w.win:add_signal("key-press", function (_, mods, key, send)
+            -- avoid double mapping
+            if send then
```
this should be all one line; but the real problem is that this breaks people using external tools to automate luakit (is this likely? no, but it's still an API break)

I think the best solution for now is to add a `window.act_on_synthetic_keys` setting (feel free to think of a nicer name), and add it to the if-condition.

aidanholm commented on widgets/common.c:
```diff
@@ -33,7 +33,8 @@ key_press_cb(GtkWidget* UNUSED(win), GdkEventKey *ev, widget_t *w)
     luaH_modifier_table_push(L, ev->state);
     luaH_keystr_push(L, ev->keyval);
-    gint ret = luaH_object_emit_signal(L, -3, "key-press", 2, 1);
+    lua_pushboolean(L, ev->send_event);
```
:+1:

aidanholm commented on widgets/window.c:
```diff
@@ -71,6 +72,83 @@ luaH_window_set_default_size(lua_State *L)
+    luaH_checktable(L, 3);
+
+    lua_pushnil(L);
+    while(lua_next(L, 3)) {
```
I'm not happy with this bind syntax being different to the bind syntax used for keybindings. It might be a bit of work unifying that, but I'd also rather not get locked into having two syntaxes for compatibility reasons...

aidanholm commented on widgets/window.c:
```diff
@@ -71,6 +72,83 @@ luaH_window_set_default_size(lua_State *L)
+    GdkKeymapKey* keys = NULL;
+    gint n_keys;
+
+    if (!gdk_keymap_get_entries_for_keyval(gdk_keymap_get_default(), keyval, &keys, &n_keys)) {
```
`gdk_keymap_get_default()` should be `NULL`

amosbird replied:
Doesn't work if I change it to `NULL`

aidanholm commented on widgets/window.c:
```diff
@@ -71,6 +72,83 @@ luaH_window_set_default_size(lua_State *L)
+        return 1;
+    }
+
+    g_assert(keys);
```
can probably get rid of this as well

aidanholm commented on widgets/window.c:
```diff
@@ -71,6 +72,83 @@ luaH_window_set_default_size(lua_State *L)
+
+    g_free(keys);
+
+    gdk_event_put(event);
```
I'd say move this `gdk_event_put()` call up against the `event_key` block above

aidanholm commented on widgets/window.c:
```diff
@@ -71,6 +72,83 @@ luaH_window_set_default_size(lua_State *L)
 }
 
 static gint
+luaH_window_send_key(lua_State *L)
```
I haven't really looked at this code, but it's spamming warnings and that's a no no
```
Gdk-WARNING **: Event with type 8 not holding a GdkDevice. It is most likely synthesized outside Gdk/GTK+
```

aidanholm replied:
Also, in general this could do with some logging: probably just a `debug("sending key '%s' to window %p", ...)` is sufficient. And the validation should use `luaL_error()` on failure: this is a function that only fails if the user screwed up the argument.

aidanholm commented on config/rc.lua:
```diff
@@ -61,6 +61,9 @@ require "settings_chrome"
 local adblock = require "adblock"
 local adblock_chrome = require "adblock_chrome"
 
+-- Add ability to send keystrokes to the window
```
Do we want this enabled by default? I'm undecided...

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,71 @@
+    local modifiers = {}
+    local keys = {}
+    -- https://www.lua.org/manual/5.3/manual.html#pdf-utf8.codes
+    for char in string.gmatch(keystring, "[%z\1-\x7F\xC2-\xF4][\x80-\xBF]*") do
```
as mentioned on IRC, ugh; I'll have a look at getting `utf8.codes` in

aidanholm commented:
In addition, there are two test failures:
```
FAIL style/test_source_format / test_vim_modeline
  tests/style/test_source_format.lua:41: Some files do not have modelines:
  lib/keysym.lua                                         Missing/malformed modeline
FAIL style/test_source_format / test_lua_module_functions_are_documented
  tests/style/test_source_format.lua:262: Some Lua modules have documentation issues:
  lib/keysym.lua:22                                      Undocumented export 'send'
```
The modeline is the `-- vim: et:sw=4:ts=8:sts=4:tw=80` thing that every file has. The new window `send_keys` method will need documenting too, but that's more of a final task.

aidanholm commented:
Oh, and another thing I'd prefer is if you could add your changes on top of the existing commits in the keysym branch, as that preserves history: both the reasoning behind each section of code, and the attribution to the original authors

amosbird committed:
Fake keys in luakit!

This patch is a rebase of the keysym branches, with some improvements and tenderly care of @aidanholm :)

aidanholm commented on lib/window.lua:
```diff
@@ -709,6 +712,11 @@ function _M.ancestor(w)
+    ["window.act_on_synthetic_keys"] = {
+        type = "boolean",
+        default = false,
+        desc = "Whether or not to act on syntheic keys.",
```
At the moment this is just restating the setting name; it should be more descriptive, e.g: "Whether synthetic key events should activate key bindings. Synthetic key events have been generated by a program rather than a physical key press, such as those sent by `keysym.send()`."

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,72 @@
+    local symbol = nil
+    local modifiers = {}
+    local keys = {}
+    for char in string.gmatch(keystring, utf8.charpattern) do
```
The usual style here is to use method access notation: `keystring:gmatch(utf8.charpattern)`

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,72 @@
+-- @tparam w The window object.
+-- @tparam string keystring The key string representing synthetic keys.
+_M.send = function (w, keystring)
+    local symbol = nil
```
this needs assertions that `w` is a table and `keystring` is a string

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,72 @@
+            symbol = nil
+            modifiers = {}
+        elseif symbol and char == "-" then
+            if string.match(symbol, "^[Ss]hift$") or
```
ditto

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,72 @@
+            symbol = symbol .. char
+        end
+    end
+    assert(not symbol)
```
This needs an "unterminated symbol" message

aidanholm commented on widgets/window.c:
```diff
@@ -90,8 +89,7 @@ luaH_window_send_key(lua_State *L)
     }
 
     if (!keyval || keyval == GDK_KEY_VoidSymbol) {
```
No need for braces with a single line body

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,78 @@ luaH_window_set_default_size(lua_State *L)
+    }
+
+    guint state = 0;
+    luaH_checktable(L, 3);
```
The checktable call should be moved to the top with the other argument checks

aidanholm replied:
Actually, the table should be optional, but if there is a third argument it must be a table

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,78 @@ luaH_window_set_default_size(lua_State *L)
+    event_key->hardware_keycode = keys[0].keycode;
+    event_key->group = keys[0].group;
+    gdk_event_put(event);
+    debug("sending key '%s' with modifier %d to window %p",
```
The window pointer printed should be the widget (`w`)

aidanholm reviewed (commented):
Just a few minor fixes needed

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
+{
+    widget_t *w = luaH_checkwidget(L, 1);
+    const gchar *key_name = luaL_checkstring(L, 2);
+    luaH_checktable(L, 3);
```
This needs to be inside `if (lua_gettop(L) >= 3)`, as we want to allow leaving the table out.

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
+{
+    widget_t *w = luaH_checkwidget(L, 1);
+    const gchar *key_name = luaL_checkstring(L, 2);
+    luaH_checktable(L, 3);
```
This needs to be inside `if (lua_gettop(L) >= 3)`, as we want to allow leaving the table out.

aidanholm replied:
in order not to add another check below in the modifier loop, you could add an `else lua_newtable(L);` to that

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
+    luaH_checktable(L, 3);
+
+    if (!g_utf8_validate(key_name, -1, NULL))
+        return luaL_error(L, "key name isn't a utf-8 string.");
```
no full stop on error messages

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
+
+    GdkKeymapKey* keys = NULL;
+    gint n_keys;
+
```
no gap here

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
+    event_key->keyval = keyval;
+    event_key->hardware_keycode = keys[0].keycode;
+    event_key->group = keys[0].group;
+    gdk_event_set_device(event, gdk_seat_get_keyboard(gdk_display_get_default_seat(gdk_display_get_default())));
```
super long line here; should wrap it at the least, or break it into a `GdkSeat *seat = ...` line above it.

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
+        lua_pop(L, 1);
+    }
+
+    GdkKeymapKey* keys = NULL;
```
\* goes next to the variable name: 'GdkKeymapKey *keys`

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
+    guint state = 0;
+    GString *state_string = g_string_sized_new(32);
+    lua_pushnil(L);
+    while(lua_next(L, 3)) {
```
space between `while` and `(`

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
+
+    g_string_free(state_string, TRUE);
+    g_free(keys);
+    return 1;
```
should be returning 0; this function doesn't return a lua value

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,74 @@
+        elseif char == ">" and symbol then
+            if #symbol == 0 then
+                error("bad keystring: " .. keystring)
+                return
```
no need for `return` here: `error()` unwinds the stack

aidanholm commented on lib/keysym.lua:
```diff
@@ -0,0 +1,74 @@
+
+local _M = {}
+
+--- Send synthetic keys to given window.
```
This needs more documentation of what sending keys means, and what the keystring format is. Ideally, a few examples, involving tricky cases like sending `<` and `>` keys, keys with modifiers, unicode keys..

aidanholm commented on widgets/window.c:
```diff
@@ -70,6 +71,76 @@ luaH_window_set_default_size(lua_State *L)
 }
 
+static gint
+luaH_window_send_key(lua_State *L)
```
this needs some documentation added to `doc/luadoc/widget.window.lua`; can be a modified version of the keysym documentation

aidanholm reviewed (changes requested):
more minor changes needed: mostly code style and documentation

aidanholm commented on widgets/window.c:
```diff
@@ -76,60 +76,52 @@ luaH_window_send_key(lua_State *L)
 {
     widget_t *w = luaH_checkwidget(L, 1);
     const gchar *key_name = luaL_checkstring(L, 2);
+    if (lua_gettop(L) >= 3) luaH_checktable(L, 3);
```
should have a newline after `if (...)` and `else`

aidanholm commented on widgets/window.c:
```diff
@@ -76,60 +76,52 @@ luaH_window_send_key(lua_State *L)
-    g_assert(keys);
+    if (!gdk_keymap_get_entries_for_keyval(gdk_keymap_get_default(),
+                                           keyval, &keys, &n_keys))
+        return luaL_error(L, "failed to get key map entries.");
```
no fullstop here

aidanholm commented on doc/luadoc/widget.window.lua:
```diff
@@ -15,6 +15,15 @@
 -- @tparam integer height The default height of the window, in pixels.
 
+--- @method send_key
+-- Send synthetic keys to given window. This function parses a vim-like
```
s/synthetic keys/synthetic key events/

aidanholm commented on doc/luadoc/widget.window.lua:
```diff
@@ -15,6 +15,15 @@
+--- @method send_key
+-- Send synthetic keys to given window. This function parses a vim-like
+-- keystring into single keys and sends them to the window. When
+-- `window.act_on_synthetic_keys` is on, synthetic key events will not trigger
```
s/on/disabled/

aidanholm commented on lib/keysym.lua:
```diff
@@ -1,29 +1,48 @@
+--     keysym.send(w, "<Control-a>")
+--     keysym.send(w, "世界，你好")
+--
+-- When `window.act_on_synthetic_keys` is on, synthetic key events will
```
s/on/disabled/ here as well

aidanholm reviewed (changes requested):
few more minor fixes needed

aidanholm commented on widgets/window.c:
```diff
@@ -76,60 +76,55 @@ luaH_window_send_key(lua_State *L)
-    g_assert(keys);
+    if (!gdk_keymap_get_entries_for_keyval(gdk_keymap_get_default(),
+                                           keyval, &keys, &n_keys))
+        return luaL_error(L, "cannot type '%s' on current keyboard layout",
```
need to free `state_string` here as well, otherwise we leak memory
